### PR TITLE
BAAS-22402: add early exits when possible memusage

### PR DESCRIPTION
**Pull request by nickpoindexter**
In this PR it seems like we have opportunity to exit large arrays early when we've exceeded the total mem allowed. This does not cause the interrupt in the function and will continue to rely on the poller within pass to check these values but should ease the CPU needed to calculate mem.

I think this also allows us to remove the estimate, but that path is a bit different so was not adjusted as part of this effort.

### CODE REVIEW

File: .evg.yml [hunk 1]
```diff
@@ -25,7 +25,11 @@ tasks:
             export PATH=$ROOT_DIR/go/bin:$PATH
 
             cd goja
-            go test -v ./...
+            go test -v ./... >> ${workdir}/goja.suite
+      - command: gotest.parse_files
```
Comment (nickpoindexter) [inline .evg.yml:29]: [drive-by] to get tests under tests tab in EVG

File: memory_test.go [hunk 2]
```diff
@@ -213,28 +213,28 @@ func TestMemCheck(t *testing.T) {
 			// len("abc") + len("a") + SizeNumber
 			3 + 1 + SizeNumber,
 		},
-		{
```
Comment (nickpoindexter) [inline memory_test.go:216]: Gabri is adjusting mem tracking tests in his current work. Since these were adjusted with the early cutoff, I have just commented these out.

File: memory_test.go [hunk 3]
```diff
@@ -521,22 +521,22 @@ func TestMemArraysWithLenThreshold(t *testing.T) {
...
-			};
-			checkMem()`,
-			200,
-			40,
-			// Array overhead, size of property values, only 3 values before we hit the mem limit
-			2 + SizeEmpty + (3 * SizeBool),
-		},
+			SizeEmpty + (50 * SizeNumber) + (50 * SizeBool),
+		},
+		// {
```
Comment (nickpoindexter) [inline memory_test.go:526]: Gabri is adjusting mem tracking tests in his current work. Since these were adjusted with the early cutoff, I have just commented these out.

File: memory_test.go [hunk 4]
```diff
@@ -625,21 +625,21 @@ func TestMemObjectsWithPropsLenThreshold(t *testing.T) {
...
-				}
-				checkMem()`,
-			100,
-			40,
-			// len("i0") + value i
-			10*2 + 10*SizeNumber +
-				// for loop 0 + 10 + len("i")
-				SizeNumber + SizeNumber + 1,
-		},
+		// {
```
Comment (nickpoindexter) [inline memory_test.go:628]: Gabri is adjusting mem tracking tests in his current work. Since these were adjusted with the early cutoff, I have just commented these out.

File: array_sparse_test.go [hunk 5]
```diff
@@ -262,3 +262,104 @@ func TestSparseArrayExportToSlice(t *testing.T) {
 		}
 	}
 }
+
+func TestSparseArrayObjectMemUsage(t *testing.T) {
+
```
Comment (Gabri3l) [inline array_sparse_test.go:267]: [nit] empty line

File: builtin_map_test.go [hunk 6]
```diff
@@ -242,3 +242,102 @@ func BenchmarkMapDeleteJS(b *testing.B) {
 		}
 	}
 }
+
+func TestMapObjectMemUsage(t *testing.T) {
+
```
Comment (Gabri3l) [inline builtin_map_test.go:247]: [nit] empty line

File: compiler_test.go [hunk 7]
```diff
@@ -5645,3 +5646,91 @@ func BenchmarkCompile(b *testing.B) {
 		}
 	}
 }
+
+func TestProgramMemUsage(t *testing.T) {
+
```
Comment (Gabri3l) [inline compiler_test.go:5651]: [nit] empty line

File: memory_test.go (hunk 4 above)
Comment (Gabri3l) [inline memory_test.go:628]: [nit] You can undo this change now

**Review by Gabri3l — CHANGES_REQUESTED** [review]
Thank you for these changes, were you able to see what impact they end up having?

File: array_test.go [hunk 8]
```diff
@@ -131,3 +131,80 @@ func BenchmarkArraySetEmpty(b *testing.B) {
...
+				},
+			},
+			ao: &arrayObject{
+				values: []Value{
+					New()._newString(newStringValue("key"), nil),
+				},
+			},
+			expected:    41,
+			errExpected: errMemUsageExceedsLimitNil,
+		},
```
[the remark this answers is not in the document]
Comment (nickpoindexter) [reply array_test.go:192]: ahh nice catch. I added those in.

File: mem_context.go [hunk 9]
```diff
@@ -122,6 +122,20 @@ func NewMemUsageContext(
...
 }
 
+// MemUsageLimitExceeded ensures a limit function is defined and checks against the limit. If limit is breached
+// it will return true, if the limit check is undefined it will return an error
+func (m *MemUsageContext) MemUsageLimitExceeded(memUsage uint64) (bool, error) {
+	if m.MemUsageExceedsLimit == nil {
+		return false, errMemUsageExceedsLimitNil
+	}
+	// This is an early exit in case we reach the mem usage
+	// limit before we get to scan the whole array.
```
Comment (tcannon91) [inline mem_context.go:132]: [nit] This comment seems applicable to array use cases but not the map or compiler...
```suggestion
```

**Review by tcannon91 — APPROVED** [review]
LGTM

File: mem_context.go [hunk 10]
```diff
@@ -122,6 +122,18 @@ func NewMemUsageContext(
 	}
 }
 
+// MemUsageLimitExceeded ensures a limit function is defined and checks against the limit. If limit is breached
+// it will return true, if the limit check is undefined it will return an error
+func (m *MemUsageContext) MemUsageLimitExceeded(memUsage uint64) (bool, error) {
+	if m.MemUsageExceedsLimit == nil {
```
Comment (arahmanan) [inline mem_context.go:128]: do we need to perform this check in tight loops? I would rather not if possible. Wouldn't we be able to check this in advance?

Comment (nickpoindexter) [reply mem_context.go:128]: The var is an exported pointer structure so anyone could change it at any point. I think if we determined that the only ever definition of this function would be:
```go
if memUsage > m.MemLimit {
return err
}
```
then we can remove these checks and clean things up a bit. Happy to do so if everyone is on board.

Comment (nickpoindexter) [reply mem_context.go]: Chatted with @Gabri3l offline and he was fine with moving forward this approach as long as the limit field was not exported. It has been refactored.

**Review by Gabri3l — APPROVED** [review]
Looking good 💪🏼

**Review by arahmanan — APPROVED** [review]
🥳
